### PR TITLE
Cache `Circuit` properties between mutations

### PR DESCRIPTION
**Pull request by maffoo**
This simplifies working with frozen circuits and circuit identity by storing the `FrozenCircuit` instance returned by `Circuit.freeze()` and returning the same instance until it is "invalidated" by any mutations of the Circuit itself. Note that if we make additional changes to the Circuit implementation, such as adding new mutating methods, we will need to remember to add `self._frozen = None` statements to invalidate the frozen representation in those places as well. To make this easier to audit, I have put these invalidation statements immediately after any mutation operations on `self._moments`, even in places where some invalidations could be elided or pushed to the end of a method; it seems safer to keep these close together in the source code. I have also added an implementation note calling out this detail and reminding future developers to invalidate when needed if adding other `Circuit` mutations.

### CODE REVIEW

**Comment by maffoo** [comment]
Something to note is that there are a number of other circuit methods that could be optimized in the same way. For example, `Circuit.all_qubits()` requires traversing all operations in the circuit to compute the value, but we should be able to cache it until the circuit is mutated. This would improve performance in lots of real-world scenarios.

**Comment by maffoo** [comment]
I updated this to cache various other circuit parameters as well. To avoid slowing down mutation methods too much, I've changed this to call the `self._mutated()` method which resets cached properties only once in any given mutating method.

File: cirq-core/cirq/circuits/circuit.py [hunk 1]
```diff
@@ -1769,6 +1773,12 @@ def __init__(
             else:
                 self.append(flattened_contents, strategy=strategy)
 
+    def _mutated(self) -> None:
```
Comment (NoureldinYosri) [inline cirq-core/cirq/circuits/circuit.py:1776]: shouldn't this also include `_is_measurement` and `_parameter_names`?

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]
Overall looks good, but I think we need tests of the form
```
for each property X that we cache
    query property X -> which caches the result
    for each method Y that mutates X
        do Y
        query X again and check that the result is correct
```

File: cirq-core/cirq/circuits/circuit.py (hunk 1 above)
Comment (maffoo) [reply cirq-core/cirq/circuits/circuit.py:1776]: Good catch. Fixed.

File: cirq-core/cirq/circuits/circuit.py [hunk 2]
```diff
@@ -1831,6 +1843,41 @@ def _load_contents_with_earliest_strategy(self, contents: 'cirq.OP_TREE'):
     def __copy__(self) -> 'cirq.Circuit':
         return self.copy()
 
+    def freeze(self) -> 'cirq.FrozenCircuit':
+        """Gets a frozen version of this circuit.
+
+        Repeated calls to `.freeze()` will return the same FrozenCircuit
+        instance as long as this circuit is not mutated.
+        """
+        from cirq.circuits import FrozenCircuit
```
Comment (NoureldinYosri) [inline cirq-core/cirq/circuits/circuit.py:1852]: nit: should this be a full import path `cirq.circuits.frozen_circuit`?

**Review by NoureldinYosri — APPROVED** [review]

File: cirq-core/cirq/circuits/circuit.py (hunk 2 above)
Comment (maffoo) [reply cirq-core/cirq/circuits/circuit.py:1852]: This was copied from what we had in `AbstractCircuit.freeze` previously. I don't feel strongly either way, though. Will change to use full import path.

**Pull request merged**